### PR TITLE
amcvideodec: don't send eos on flow error

### PR DESCRIPTION
**Pull request by aslobodeniuk**
It was causing next deadlock:
gst_pad_push_event can never return, because
up the stack there will be a sink waiting
for a preroll that will never happen.

### CODE REVIEW

File: sys/androidmedia/gstamcvideodec.c [hunk 1]
```diff
@@ -1363,9 +1363,12 @@ gst_amc_video_dec_loop (GstAmcVideoDec * self)
   /* We're going to stop srcpad's loop until new buffers on sinkpad */
 
   if (error_msg) {
+    GST_ERROR_OBJECT (self, "%s", error_msg);
```
Comment (dabrain34) [inline sys/androidmedia/gstamcvideodec.c]: No need for this log, the element error is enough

**Review by dabrain34 — APPROVED** [review]

File: sys/androidmedia/gstamcvideodec.c [hunk 2]
```diff
@@ -1363,9 +1363,12 @@ gst_amc_video_dec_loop (GstAmcVideoDec * self)
...
     GST_ELEMENT_ERROR (self, LIBRARY, FAILED, (NULL), ("### %s", error_msg));
-    gst_pad_push_event (GST_VIDEO_DECODER_SRC_PAD (self), gst_event_new_eos ());
     self->downstream_flow_ret = GST_FLOW_ERROR;
```
Comment (dabrain34) [inline sys/androidmedia/gstamcvideodec.c:1367]: why did you change the order with self->downstream_flow_ret

**Review by dabrain34 — DISMISSED** [review]

File: sys/androidmedia/gstamcvideodec.c (hunk 2 above)
Comment (aslobodeniuk) [reply sys/androidmedia/gstamcvideodec.c:1367]: For us it's important to set this variable before unlocking the stream.
Since we're unlocking the stream, another thread that is checking self->downstream_flow_ret is allowed to run.

Comment (aslobodeniuk) [reply sys/androidmedia/gstamcvideodec.c:1367]: order with eos downstream doesn't matter

Comment (turran) [reply sys/androidmedia/gstamcvideodec.c:1367]: I dont like this approach. Is like leaving a new time window to let the thread that pushes data into the decoder run, when there's no real need. The STREAM_LOCK must be held while doing stuff on the streaming thread, like the case of pushing an EOS. It is on purpose to avoid the input thread push data into the decoder.

The EOS is a "prerollable" object, check the docs at https://www.freedesktop.org/software/gstreamer-sdk/data/docs/latest/gstreamer-libs-0.10/GstBaseSink.html specially the part: " When the element receives EOS in PAUSED, preroll completes, the event is queued and an EOS message is posted when going to PLAYING."

So, even if the error arrives to the upper layer and wants to send the element to ready it should have the STREAM_LOCK unlocked already, unless there's another bug and your fix is just hiding it.

Comment (aslobodeniuk) [reply sys/androidmedia/gstamcvideodec.c:1367]: Well yes, in general you're right. I also was not sure about a solution, but only that it will not break anything.

> Is like leaving a new time window to let the thread that pushes data into the decoder run, when there's no real need

streaming thread will not push data into decoder because of
self->downstream_flow_ret = GST_FLOW_ERROR
it will immediatelly quit chain function with GST_FLOW_ERROR

> When the element receives EOS in PAUSED, preroll completes, the event is queued and an EOS message is posted when going to PLAYING.

Basesink didn't reach PAUSED state, decoder is pushing an event from a typefind. Maybe this is the reason of a deadlock: pushing EOS before any output.

Also another possible reason is that from the code of amcvideodec we may push the eos event to the pad that before returned a bad state to us.

Actually I don't see a reason to push eos in error case at all, if we've already emited an error message, maybe you can tell me.
I could just remove this "eos in error case", to avoid unexpected behaviour in fluendo-sdk. FluVAdec is not doing this.

Comment (turran) [reply sys/androidmedia/gstamcvideodec.c:1367]: Ok, just saw it. As i was saying before the EOS is a prerolled element, so if there are elements in between the decoder and the sink that EOS should be pushed also by the in-the-middle elements. I would say that this needs further investigation. We can skip sending the EOS on the decoder given that we are also triggering the element error and that is going to remove the branch.

**Review by turran — APPROVED** [review]
